packages/calypso-css-custom-properties/src/shared/_colors.scss

```diff
@@ -1,5 +1,5 @@
 // The Muriel Color Palette
-@import '~color-studio/dist/color-variables';
+@import '../../node_modules/color-studio/dist/color-variables';
```

Comment:
I essentially reverted [this change](https://github.com/Automattic/wp-calypso/pull/30969/files#diff-82edfa02ea2dc5c3903ce359579307d9R2) here. The reason is that `node-sass-package-importer` isn't great at transitivity:

Calypso doesn't just import `calypso-css-custom-properties`' [main entry point](https://github.com/Automattic/wp-calypso/pull/30933/files#diff-6719378e57fed866cca7f644c57674eaR25) (`dist/custom-properties.css`), but continues to also import this file (`shared/colors`) which in turn has the import of `color-studio`'s `color-variables` file I'm modifying here.

Now if we used the `node-sass-package-importer`-based import notation (`@import '~color-studio/dist/color-variables'`), any consumer of that package would look for that path in its _own_ `node_modules/` directory -- rather than (transitively) in `calypso-sdk-custom-properties`. It follows that:

- Any consumer of `calypso-sdk-custom-properties` (including Calypso!) will either have to install `color-studio` itself (by adding it to its own `package.json`, or
- `calypso-css-custom-properties` would need to add _another_ build step to provide a file where this `import` is resolved.

Or we just make the `import1 path explicit like I'm doing here, which seems like a lot less hassle :smile: it's `calypso-css-custom-properties`'s only import of an external SASS file anyway.

Reply:
I think you _have_ to use the tilde syntax here. The location of `node_modules` relative to this file isn't guaranteed.

Reply:
Hmm, sure about that? (Or maybe I'm really getting my wires crossed here.)

This (i.e. the `src/shared/_colors.scss` file in the `calypso-css-custom-properties` package) is referring to that package's own `node_modules/` directory here (whose relative location should thus be known and invariant).

The issue is with consuming projects that import this file: if they see a `@import '~color-studio/dist/color-variables'` in it, they'll look for that file in their _own_ `node_modules/` rather than in `calypso-css-custom-properties`'s -- and they won't fit there (unless it's also installed there -- but requiring them to would kinda suck, see my previous comment).

(This is kinda backed up by my experiments with both notations :grimacing: )

Reply:
(My generalized takeaway would BTW be -- _Don't use `~` SASS imports in npm modules (libraries) but only in apps that aren't published/imported by other projects (b/c of lack of transitivity)_)

Reply:
>  This (i.e. the `src/shared/_colors.scss` file in the `calypso-css-custom-properties` package) is referring to that package's own `node_modules/` directory here (whose relative location should thus be known and invariant).

Try running `npm dedupe` on this branch. `color-studio` will get hoisted to the root `node_modules` and the `node_modules` under `color-studio` will be removed. Yay for npm's non-determinism. Another way to get there would be to add `color-studio` as a dependency to the root package.json, then run `npm run update-deps`.


> The issue is with consuming projects that import this file: if they see a `@import '~color-studio/dist/color-variables'` in it, they'll look for that file in their _own_ `node_modules/` rather than in `calypso-css-custom-properties`'s

Consuming projects should depend on the package, which depends on color-studio, so it should be present somewhere in the node_modules tree. If consuming projects are using this file on it's own without the supporting npm infrastructure ... I dunno.

Reply:
> My generalized takeaway would BTW be -- _Don't use `~` SASS imports in npm modules (libraries) but only in apps that aren't published/imported by other projects (b/c of lack of transitivity)_

If we do expect folks to import the SASS directly, we'll also have to document that they need to also use the fancy importer plugin that understands the ~. Or webpack, which it emulates.

Reply:
> > This (i.e. the `src/shared/_colors.scss` file in the `calypso-css-custom-properties` package) is referring to that package's own `node_modules/` directory here (whose relative location should thus be known and invariant).
> 
> Try running `npm dedupe` on this branch. `color-studio` will get hoisted to the root `node_modules` and the `node_modules` under `color-studio` will be removed. 

Okay, can repro -- that makes the build fail :slightly_frowning_face: 

> Yay for npm's non-determinism. Another way to get there would be to add `color-studio` as a dependency to the root package.json, then run `npm run update-deps`.
> 
> > The issue is with consuming projects that import this file: if they see a `@import '~color-studio/dist/color-variables'` in it, they'll look for that file in their _own_ `node_modules/` rather than in `calypso-css-custom-properties`'s
> 
> Consuming projects should depend on the package, which depends on color-studio, so it should be present somewhere in the node_modules tree. If consuming projects are using this file on it's own without the supporting npm infrastructure ... I dunno.

So to recap: Starting from a clean slate (`npm run distclean && npm ci`):

- With 
`@import '../../node_modules/color-studio/dist/color-variables';`: after `npm dedupe`, `npm start` fails because `color-studio` has been hoisted to the top-level `node_modules/`. (It works if `npm dedupe` _isn't_ called, though.)
- With `
@import '~color-studio/dist/color-variables';`: `npm start` fails, because it's looking for `color-studio` in the top-level `node_modules/` (but it hasn't been installed explicitly there, nor hoisted. `npm dedupe` makes it work).

So how do we get out of this mess? Just install `color-studio` explicitly at the top level?

Reply:
> With ` @import '~color-studio/dist/color-variables';`: `npm start` fails, because it's looking for `color-studio` in the top-level `node_modules/` (but it hasn't been installed explicitly there, nor hoisted. `npm dedupe` makes it work).

In this case, who's doing the import? Something in assets/stylesheets or something in calypso-css-custom-properties? If the former, yeah, we'd have to add a dependency on color-studio to the root for resolution to work. If the latter, ... I would expect that to work and would need to dig more.

Reply:
That's an interesting observation. So, with the following directory structure:
```
node_modules/color-studio/index.scss
color-schemes/node_modules/color-studio/index.scss
color-schemes/index.scss
```
and contents of `color-schemes/index.scss`:
```
@import '~color-studio';
```
the import gets resolved to `node_modules/color-studio` instead of `color-schemes/node_modules/color-studio`? That's not how the resolution algorithm should work and it's a bug in `node-sass-package-importer`.

The `node-sass` importer function gets the file path from where the import is performed as the second parameter and it should use it to resolve:
```js
function importer( url, prev ) {
  if ( ! url.startsWith( '~' ) ) {
    return null; // not our business to resolve this path
  }
  return resolve.sync( url.slice( 1 ), { baseDir: dirname( prev ) } );
}
```
What `node-sass-package-importer` does instead is that is passes `process.cwd()` as `baseDir`.

The other package that @blowery suggested in https://github.com/Automattic/wp-calypso/pull/30933#discussion_r259107727, https://github.com/anarh/node-sass-import, doesn't have this bug. But it doesn't support the `~` prefix 🙁 

Not a reason to block this PR, the `__color-studio` solution is OK, but would be nice to file some issues and come back some time later.

Reply:
Thanks for digging into that @jsnajdr !

Reply:
Filed https://github.com/maoberlehner/node-sass-magic-importer/issues/188